libs/go/aealite/protocols/dialogue.go

```diff
@@ -0,0 +1,437 @@
+package protocols
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+)
+
+type Address string
+type Performative []string
+type SomeMessageType string
+
+const (
+	NONCE_BYTES_NB = 32
+)
+
+type Role string
+
+const (
+	Role1             Role = "role1"
+	Role2             Role = "role2"
+	StartingMessageId      = 1
+	StartingTarget         = 0
+)
+
+type RuleType struct {
+	initialPerformatives  Performative
+	terminalPerformatives Performative
+	validReplies map[Performative][]Performative
+}
+
+type MessageId int
+type Target int
+type DialogueInterface interface {
+	getSelfAddress() Address
+	getRole() Role
+	getRules() RuleType
+	getMessageClass() AbstractMessage
+	isSelfInitiated() bool
+	getLastIncomingMessage() AbstractMessage
+	getLastOutgoingMessage() AbstractMessage
+	getLastMessage() AbstractMessage
+	counterPartyForMessage(AbstractMessage) bool
+	isMessageBySelf(AbstractMessage) bool
+	isMessageByOther(AbstractMessage) bool
+	getMessage(MessageId) AbstractMessage
+	hasMessageId(MessageId) bool
+
+	update(AbstractMessage)
+	validateNextMessage(AbstractMessage) (bool, string)
+	basicValidations(AbstractMessage) (bool, string)
+	basicValidationInitialMessage(AbstractMessage) (bool, string)
+	basicValidationNonInitialMessage(AbstractMessage) (bool, string)
+	isEmpty() bool
+	updateIncomingAndOutgoingMessages(AbstractMessage)
+	isBelongingToADialogue(AbstractMessage) bool
+
+	validateMessageTarget(AbstractMessage) string
+	validateMessageId(AbstractMessage) string
+	getMessageById(MessageId) AbstractMessage
+	getOutgoingNextMessageId() int
+	getIncomingNextMessageId() int
+	updateDialogueLabel(DialogueLabel)
+	customValidation(AbstractMessage) (bool, string)
+	getStringRepresentation() string
+}
+type AbstractMessageInterface interface {
+	validPerformatives() []string
+	hasSender() (Address, error)
+	hasCounterparty() (Address, error)
+	dialogueReference() [2]string
+	messageId() MessageId
+	performative() Performative
+	target() Target
+}
+
+func (message AbstractMessage) hasSender() (Address, error) {
+	if message.sender != "" {
+		return message.sender, nil
+	}
+	return "", errors.New("Sender address does not exist.")
```

Comment:
(disclaimer: very picky comment)

> Error strings should not be capitalized (unless beginning with proper nouns or acronyms) or end with punctuation, since they are usually printed following other context.
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
feel free to _not_ follow this advice :slightly_smiling_face: